src/overlays/actors/ovl_Arrow_Fire/z_arrow_fire.c

```diff
@@ -21,21 +33,214 @@ const ActorInit Arrow_Fire_InitVars = {
     (ActorFunc)ArrowFire_Update,
     (ActorFunc)ArrowFire_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arrow_Fire/func_80865D10.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arrow_Fire/ArrowFire_Init.s")
+static InitChainEntry initChain[] = {
+    ICHAIN_F32(unk_F4, 2000, ICHAIN_STOP),
+};
+
+void ArrowFire_SetupAction(ArrowFire* this, ActorFunc* actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void ArrowFire_Init(ArrowFire* this, GlobalContext* globalCtx) {
+    Actor_ProcessInitChain(&this->actor, initChain);
+    this->radius = 0;
+    this->unk_158 = 1.0f;
+    ArrowFire_SetupAction(&this->actor, ArrowFire_Charge);
+    Actor_SetScale(this, 0.01);
+    this->alpha = 0xA0;
+    this->timer = 0;
+    this->unk_15C = 0.0f;
+}
+
+void ArrowFire_Destroy(ArrowFire* this, GlobalContext* globalCtx) {
+    func_800876C8(globalCtx);
+    LogUtils_LogThreadId("../z_arrow_fire.c", 421);
+    // Translates to: ""Disappearance" = Disappearance"
+    osSyncPrintf("\"消滅\" = %s\n", "消滅");
+}
+
+void ArrowFire_Charge(ArrowFire* this, GlobalContext* globalCtx) {
+    EnArrow* arrow;
+
+    arrow = this->actor.attachedA;
+    if ((arrow == NULL) || (arrow->actor.update == NULL)) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+
+    if (this->radius < 10) {
+        this->radius += 1;
+    }
+    // copy position and rotation from the attached arrow
+    this->actor.posRot.pos = arrow->actor.posRot.pos;
+    this->actor.shape.rot = arrow->actor.shape.rot;
+
+    func_8002F974(&this->actor, NA_SE_PL_ARROW_CHARGE_FIRE);
+
+    // If arrow's attached is null, Link has fired the arrow
+    if (arrow->actor.attachedA == NULL) {
+        this->unkPos = this->actor.posRot.pos;
+        this->radius = 10;
+        ArrowFire_SetupAction(this, ArrowFire_Fly);
+        this->alpha = 0xFF;
+    }
+}
+
+void func_80865ECC(Vec3f* unkPos, Vec3f* firePos, f32 scale) {
+    unkPos->x += ((firePos->x - unkPos->x) * scale);
+    unkPos->y += ((firePos->y - unkPos->y) * scale);
+    unkPos->z += ((firePos->z - unkPos->z) * scale);
+}
+
+void ArrowFire_Hit(ArrowFire* this, GlobalContext* globalCtx) {
+    f32 scale;
+    f32 offset;
+    u16 timer;
+
+    if (this->actor.unk_F0 < 50.0f) {
+        scale = 10.0f;
+    } else {
+        if (950.0f < this->actor.unk_F0) {
+            scale = 310.0f;
+        } else {
+            scale = this->actor.unk_F0;
+            scale = ((scale - 50.0f) * 0.33333334f) + 10.0f;
```

Comment:
In reference to what @JoshDuMan was saying, did you try * (1.0f / 3.0f)
Not sure if that would be cleaner or not, but I've come across something similar. Or if these are used a lot as Josh had mentioned, maybe you could use a define 
#define ONE_THIRD 0.33333334f
```suggestion
            scale = ((scale - 50.0f) * ONE_THIRD) + 10.0f;
```